oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/JwtAuthenticationProvider.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * 1. If there is a "scope" or "scp" attribute, then
+ * 		if a {@link String}, then split by spaces and return, or
+ * 		if a {@link Collection}, then simply return
+ * 2. Take the resulting {@link Collection} of {@link String}s and prepend the "SCOPE_" keyword, adding
+ * 		as {@link GrantedAuthority}s.
+ *
+ * @author Josh Cummings
+ * @author Joe Grandja
+ * @since 5.1
+ * @see AuthenticationProvider
+ * @see JwtDecoder
+ */
+public final class JwtAuthenticationProvider implements AuthenticationProvider {
+	private final JwtDecoder jwtDecoder;
+
+	private static final Collection<String> WELL_KNOWN_SCOPE_ATTRIBUTE_NAMES =
+			Arrays.asList("scope", "scp");
+
+	private static final String SCOPE_AUTHORITY_PREFIX = "SCOPE_";
+
+	public JwtAuthenticationProvider(JwtDecoder jwtDecoder) {
+		Assert.notNull(jwtDecoder, "jwtDecoder cannot be null");
+
+		this.jwtDecoder = jwtDecoder;
+	}
+
+	/**
+	 * Decode and validate the
+	 * <a href="https://tools.ietf.org/html/rfc6750#section-1.2" target="_blank">Bearer Token</a>.
+	 *
+	 * @param authentication the authentication request object.
+	 *
+	 * @return A successful authentication
+	 * @throws AuthenticationException if authentication failed for some reason
+	 */
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		BearerTokenAuthenticationToken bearer = (BearerTokenAuthenticationToken) authentication;
+
+		Jwt jwt;
+		try {
+			jwt = this.jwtDecoder.decode(bearer.getToken());
+		} catch (JwtException failed) {
+			OAuth2Error invalidToken;
+			try {
+				invalidToken = invalidToken(failed.getMessage());
+			} catch ( IllegalArgumentException malformed ) {
+				// some third-party library error messages are not suitable for RFC 6750's error message charset
+				invalidToken = invalidToken("An error occurred while attempting to decode the Jwt: Invalid token");
+			}
+			throw new OAuth2AuthenticationException(invalidToken, failed);
+		}
+
+		Collection<GrantedAuthority> authorities =
+				this.getScopes(jwt)
+						.stream()
+						.map(authority -> SCOPE_AUTHORITY_PREFIX + authority)
+						.map(SimpleGrantedAuthority::new)
+						.collect(Collectors.toList());
+
+		JwtAuthenticationToken token = new JwtAuthenticationToken(jwt, authorities);
+
+		token.setDetails(bearer.getDetails());
+
+		return token;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public boolean supports(Class<?> authentication) {
+		return BearerTokenAuthenticationToken.class.isAssignableFrom(authentication);
+	}
+
+	private static OAuth2Error invalidToken(String message) {
+		return new BearerTokenError(
+				BearerTokenErrorCodes.INVALID_TOKEN,
+				HttpStatus.UNAUTHORIZED,
+				message,
+				"https://tools.ietf.org/html/rfc6750#section-3.1");
+	}
+
+	private static Collection<String> getScopes(Jwt jwt) {
+		for ( String attributeName : WELL_KNOWN_SCOPE_ATTRIBUTE_NAMES ) {
+			Object scopes = jwt.getClaims().get(attributeName);
```

Comment:
Consider using jwt.getClaimAsStringList instead.

Reply:
Actually, it appears that this method only returns the list if it is already a list, otherwise, it returns null. I'm inclined to leave this for now and submit and possibly submit an enhancement for getClaimAsStringList, maybe and overloaded method that takes a delimiter as a parameter?